### PR TITLE
Insertion sort

### PR DESCRIPTION
**Pull request by jay-tyler**

### CODE REVIEW

File: hash_table.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        self.table[self.hash(key)].append([key, val])
+
+    def hash(self, key):
+        """Return the hash value of the given key"""
+        if type(key) != str:
+            raise TypeError('Only keys of type String can be hashed.')
+        total = 0
+        for char in key:
+            total += ord(char)
+        return total % self.num_bins
```
Comment (joelstanner) [inline hash_table.py:26]: Avoid adding extraneous files to your pull requests. This shouldn't be here.

File: test_hash_table.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    with open('/usr/share/dict/words', 'r') as file:
+        lines = file.readlines()
+    lines = [line.rstrip('\n') for line in lines]
+
+    for line in lines:
+        ht.set(line, line)
+
+    for line in lines:
+        testval = ht.get(line)
+        assert testval == line
```
Comment (joelstanner) [inline test_hash_table.py:43]: Take this one out too

File: insertion_sort.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    # O(n**2) worst case
+    start = time.time()
+    for i in range(100):
+        insort(range(100)[::-1])
+    stop = time.time()
+    worst_time = (stop - start) / 100
+
+    print "Best case is {:.1f} times better than worst for n=100\n\n".format(
+        worst_time/best_time) +\
```
Comment (joelstanner) [inline insertion_sort.py:40]: Only use the backslash "\" line breaks if absolutely no other way exists to format your line breaks, and even then - try hard not to. See [here](https://google-styleguide.googlecode.com/svn/trunk/pyguide.html?showone=Line_length#Line_length) and [here](https://google-styleguide.googlecode.com/svn/trunk/pyguide.html?showone=Strings#Strings).

File: insertion_sort.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+                slist[i], slist[j] = slist[j], slist[i]
+                j -= 1
+                i -= 1
+            else:
+                break
+    return slist
+
+
+if __name__ == "__main__":
+    """Test time performance for best and worst cases"""
```
Comment (joelstanner) [inline insertion_sort.py:20]: The assignment asks that you "demonstrate the performance characteristics of this sort over a variety of lengths of input in both the best and worst-case scenarios." There is only one input length here.

File: test_insertion_sort.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    assert random_list != sorted_list
+    assert insort(random_list) == sorted_list
+
+
+def test_insertion_sort_fail():
+    for item in [1, None, (2, 1), {'1': 1, '2': 2}]:
+        with pytest.raises(TypeError):
+            insort(item)
+    with pytest.raises(TypeError):
+        insort()
```
Comment (joelstanner) [inline test_insertion_sort.py:33]: It would be good to also have some tests that test edge cases such as a zero or one item list. Also tests that ensure the sorting is working on small known lists such as [1, ..., 10] would better demonstrate the functionality of what you are trying to do. These could be in addition to your big list tests.

File: insertion_sort.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    Implementation follows after gif here
+    https://en.wikipedia.org/wiki/Insertion_sort"""
+    slist = unlist[:]
+    for i in range(1, len(slist)):
+        j = i - 1
+        while j >= 0:
+            if slist[i] < slist[j]:
+                slist[i], slist[j] = slist[j], slist[i]
+                j -= 1
+                i -= 1
```
Comment (joelstanner) [inline insertion_sort.py]: `i -= 1` isn't really doing anything here.

Comment (joelstanner) [inline insertion_sort.py:13]: You could get rid of `i -= 1` and do something like:

``` python
if slist[j] < slist[j - 1]:
    ...
```
